### PR TITLE
Generate fallback clientId for Foreign Fetch

### PR DESCRIPTION
**Pull request by jridgewell**
We use `clientId`s to distinguish between JS context windows, ensuring
that we serve one uniform AMP JS version to all requests from the
client.

Foreign Fetch, unfortunately, does not provide a `clientId`, so we must
generate our own.

This lumps all requests from a referrer into 60 second "batches",
generating a single `clientId` for those requests. After 60 seconds from
the first request, the batch's timer is invalidated and the next
request will generate a new `clientId` from the same referrer.

60 seconds was arbitrarily chosen to prevent version lock happening when
navigating back to the same page. We could have just used the referrer
as the `clientId`, but as long as the SW remains active, there would be
no way to distinguish between a navigation back. As we want to serve the
newer JS when possible, that seemed bad.

### CODE REVIEW

File: src/service-worker/core.js [hunk 1]
```diff
@@ -163,6 +163,38 @@ export function isBlacklisted(version) {
 }
 
 /**
+ * A mapping from a client's referrer into the time that referrer last made a
+ * request. This is used as a fallback to a clientId for Foreign Fetch, since
+ * it does not provide a unique clientId.
+ *
+ * @type {!Object<string, number>}
+ */
+const referrersLastRequestTime = Object.create(null);
```
Comment (cramforce) [inline src/service-worker/core.js]: Hopefully this doesn't get too big. But shouldn't in practice. Maybe leave comment to that extend.

File: src/service-worker/core.js [hunk 2]
```diff
@@ -366,7 +398,8 @@ self.addEventListener('fetch', event => {
 // Setup the Foreign Fetch listener, for when the client is on a Publisher
 // origin.
 self.addEventListener('foreignfetch', event => {
-  const response = handleFetch(event.request, event.clientId);
+  const clientId = generateFallbackClientId(event.request.referrer);
+  const response = handleFetch(event.request, clientId);
```
Comment (jridgewell) [inline src/service-worker/core.js]: Do you mean by patching the `ForeignFetchEvent` prototype? Sure.

**Comment by jridgewell** [comment]
Note this won't work well on pages with a `<meta name="referrer" content="none">` tag. Is that common?

File: src/service-worker/core.js (hunk 1 above)
Comment (jridgewell) [inline src/service-worker/core.js]: Done.

**Comment by cramforce** [comment]
`<meta name="referrer" content="none">` should be very uncommon.

File: src/service-worker/core.js (hunk 2 above)
Comment (jridgewell) [inline src/service-worker/core.js]: That seems better...

**Pull request merged**